### PR TITLE
Add functions to read and retrieve SolarAnywhere irradiance data

### PR DESCRIPTION
**Pull request by AdamRJensen**
- [x] Closes #1310
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

**Description**
This PR adds functions for reading and retrieving SolarAnywhere data as discussed in #1310. SolarAnywhere is a commercial dataset, although it provides irradiance data for a few locations for free.

**Minor comments**
The SolarAnywhere API is asynchronous, hence two requests are made for each data retrieval.

The `get_solaranywhere` function is currently not designed to handle annual and monthly data, as this seems less relevant to pvlib users.

### CODE REVIEW

**Comment by AdamRJensen** [comment]
@KWagnerCPR It would be great if you could test the functions and give some feedback on the default values.

File: docs/sphinx/source/whatsnew/v0.9.2.rst
Comment (kanderso-nrel) [inline docs/sphinx/source/whatsnew/v0.9.2.rst:16]: This will need a new [extlinks](https://www.sphinx-doc.org/en/master/usage/extensions/extlinks.html) entry for `discussion` here: 

https://github.com/pvlib/pvlib-python/blob/3f397ed44075b01f4ba535750356859195a51b2d/docs/sphinx/source/conf.py#L295-L301

Comment (AdamRJensen) [reply docs/sphinx/source/whatsnew/v0.9.2.rst:16]: Perhaps it is more appropriate to convert the discussion to an issue?

**Comment by KWagnerCPR** [comment]
> @KWagnerCPR It would be great if you could test the functions and give some feedback on the default values. 😄

@AdamRJensen Yes, I'll have a few people on our team test it out and get back to you. Which default values are you referring to?

**Comment by AdamRJensen** [comment]
@KWagnerCPR In regards to default values I'm thinking of these:

    def get_solaranywhere(latitude, longitude, api_key, start=None, end=None,
                          time_resolution=60, spatial_resolution=0.1,
                          true_dynamics=False, source='SolarAnywhereLatest',
                          variables=DEFAULT_VARIABLES, missing_data='Omit',
                          url=URL, map_variables=True, max_response_time=300)

Also, I have chosen not to parse any annual and monthly data, as that seems to be outside the scope of pvlib and would just unnecessarily complicate the function.

Furthermore, the `get_solaranywhere` function currently does not support retrieval of "probability of exceedance" time series, as the public license does not allow me to retrieve those.

**Comment by KWagnerCPR** [comment]
@AdamRJensen 

In terms of default values, I would recommend the following: 
time_resolution=60, 
spatial_resolution=0.01,
true_dynamics=False, 
source='SolarAnywhereLatest',
missing_data='FillAverage',
                    
With SolarAnywhere Public, users get access to 1 km (0.01 x 0.01 degree) spatial resolution data, so I think it makes sense to set the default as 0.01 if our focus is on time series requests rather than TGY requests. Also, using the "FillAverage" default for missing data will ensure periods of missing satellite imagery are filled with long term averages for those times, which is likely preferable. 

You should be able to request PXX data as a part of the SolarAnywhere public license. The issue might have been that you had the spatial resolution set to 0.1 when you requested it. You also have to include "ProbabilityOfExceedance": 90, in the request. 

I think it makes sense not to request the monthly and annual totals at this time.

**Comment by AdamRJensen** [comment]
@KWagnerCPR I have updated the default values accordingly. Thanks for the feedback! :smile:

Also, I figured out how to work with PXX data, so I've added a keyword called `probability_of_exceedance`.

**Comment by AdamRJensen** [comment]
@pvlib/pvlib-maintainer I am thinking of developing primarily mock tests for the `get_solaranywhere` function as the SolarAnywhere data retrieval is slow (minutes). However, I would like to include at least one real request of historical data (just two days of data, which would make the time reasonable). Thus, unless I hear any objections I will go ahead and register the pvlib email for a free SolarAnywhere account and add a GitHub Secret called `SOLARANYWHERE_API_KEY`. If there is a desire to avoid live tests at all, let me know. FYI I imagine writing mock tests for this will be a real pain.

**Comment by KWagnerCPR** [comment]
@AdamRJensen We've tested these functions and they are working well!

One thing I noticed is that when I request time series data (WeatherDataSource = SolarAnywhereLatest) without specifying a start or end time, I don't get the error message you specify in the script: "'When requesting non-TMY data, specifying `start` and 'end` is required.'". Instead I get: AttributeError: 'NoneType' object has no attribute 'tz'. Not a huge deal. Just wanted to point it out. 

I think these will be really helpful for our customers as well as new SolarAnywhere users. I'd love to include links to these in our developers documentation. Do you know if these will get documented in the official pvlib documentation?: https://pvlib-python.readthedocs.io/en/stable/index.html

**Comment by AdamRJensen** [comment]
> One thing I noticed is that when I request time series data (WeatherDataSource = SolarAnywhereLatest) without specifying a start or end time, I don't get the error message you specify in the script: "'When requesting non-TMY data, specifying `start` and 'end` is required.'". Instead I get: AttributeError: 'NoneType' object has no attribute 'tz'. Not a huge deal. Just wanted to point it out.

@KWagnerCPR Thank you for catching this error. I had forgotten to raise the error, hence nothing was happening - it was been corrected.

> I think these will be really helpful for our customers as well as new SolarAnywhere users. I'd love to include links to these in our developers documentation. Do you know if these will get documented in the official pvlib documentation?: https://pvlib-python.readthedocs.io/en/stable/index.html

Once the functions get merged and we make a new release of pvlib, the documentation will be available publicly. We're releasing version 0.9.2 within a few weeks, but I can't promise that this pull request will be merged by then.

**Comment by KWagnerCPR** [comment]
> > One thing I noticed is that when I request time series data (WeatherDataSource = SolarAnywhereLatest) without specifying a start or end time, I don't get the error message you specify in the script: "'When requesting non-TMY data, specifying `start` and 'end` is required.'". Instead I get: AttributeError: 'NoneType' object has no attribute 'tz'. Not a huge deal. Just wanted to point it out.
> 
> @KWagnerCPR Thank you for catching this error. I had forgotten to raise the error, hence nothing was happening - it was been corrected.
> 
> > I think these will be really helpful for our customers as well as new SolarAnywhere users. I'd love to include links to these in our developers documentation. Do you know if these will get documented in the official pvlib documentation?: https://pvlib-python.readthedocs.io/en/stable/index.html
> 
> Once the functions get merged and we make a new release of pvlib, the documentation will be available publicly. We're releasing version 0.9.2 within a few weeks, but I can't promise that this pull request will be merged by then.

@AdamRJensen Thank you! I'll stay tuned.

**Comment by AdamRJensen** [comment]
When downloading data from the SolarAnywhere website, missing data may either be filled with average data or represented as missing using one of three options:
image

Missing values are by default correctly converted to Numpy nan. Similarly, I belive that the 'NaN' option would also be correctly converted to Numpy nan since we're using pd.read_csv (though I haven't tested it). However, -999 will have to be converted to numpy nan by the `read_solaranywhere` function.

@KWagnerCPR Do you have a file or a know station/time period where there is missing data that I could use for testing?

**Comment by AdamRJensen** [comment]
The remote test failures are caused by this line:

    solaranywhere_api_key = os.environ["SOLARANYWHERE_API_KEY"]

which makes me assume that the API key added to the GitHub Secrets by @kanderso-nrel isn't named what I think they are...

**Comment by kanderso-nrel** [comment]
I think it is just the annoying but secure behavior of `pull-request-target` not respecting `yml` changes in the PR -- your update to include the SA key in the environment isn't taking effect because the action is defined by the `.github/workflows/pytest-remote-data.yml` file on `master`, not this PR branch.  I guess the cleanest workaround is to submit a separate PR to change `pytest-remote-data.yml`, merge that, then merge master into this PR?

**Comment by KWagnerCPR** [comment]
@cwhanse I see, thank you for sharing that link! Looks like this item was slated for the 0.10.2 release on September 6, but was then deferred. Do you have an update on when the next release may take place?

**Comment by kandersolar** [comment]
@KWagnerCPR I've added this to the v0.10.3 milestone, which is scheduled to be released mid-December.  But (as seen in the previous milestones for this PR), that's just an aspiration, not a guarantee.  Getting this PR merged depends on reviewer availability, which is in rather short supply at the moment.

**Comment by KWagnerCPR** [comment]
@kandersolar Understood. Thank you for this background info!

File: docs/sphinx/source/whatsnew/v0.9.2.rst
Comment (kandersolar) [inline docs/sphinx/source/whatsnew/v0.9.2.rst:19]: Move to 0.10.3

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [inline pvlib/iotools/solaranywhere.py:88]:
```suggestion
        requesting probability of exceedance (POE) time series. [%]
```

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [inline pvlib/iotools/solaranywhere.py:99]: Some other iotools functions have a `timeout` parameter that serves a similar purpose (although the mechanical details are a bit different under the hood).  Worth considering using the same parameter name here for consistency.

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [inline pvlib/iotools/solaranywhere.py:204]: Other iotools functions don't convert to UTC; why do it here?

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [inline pvlib/iotools/solaranywhere.py:250]: Mild preference to expose `encoding` as an optional parameter

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [inline pvlib/iotools/solaranywhere.py:229]: Should be `filename`

File: pvlib/tests/iotools/test_solaranywhere.py
Comment (kandersolar) [inline pvlib/tests/iotools/test_solaranywhere.py:202]: This comment doesn't apply to this test

File: pvlib/tests/iotools/test_solaranywhere.py
Comment (kandersolar) [inline pvlib/tests/iotools/test_solaranywhere.py:230]: Pointing out these comments so they don't get forgotten :)

**Review by kandersolar — COMMENTED** [review]
A few comments.  Generally looks good to me!

File: pvlib/iotools/solaranywhere.py
Comment (AdamRJensen) [reply pvlib/iotools/solaranywhere.py:99]: Has now been implemented

File: pvlib/iotools/solaranywhere.py
Comment (AdamRJensen) [reply pvlib/iotools/solaranywhere.py:204]: I think you are correct good sir!

File: pvlib/iotools/solaranywhere.py
Comment (AdamRJensen) [reply pvlib/iotools/solaranywhere.py:250]: Also for the ``get_solaranywhere`` function?

**Comment by AdamRJensen** [comment]
@kandersolar Could you help with the flake8 issue - I can't really tell what it is?

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [reply pvlib/iotools/solaranywhere.py]: I don't think `get_solaranywhere` has to worry about encoding problems; network requests shouldn't be subject to users unwittingly re-encoding files.

**Comment by cwhanse** [comment]
> @kandersolar Could you help with the flake8 issue - I can't really tell what it is?

Do you get the error if the csv files are read with pandas?

**Comment by kandersolar** [comment]
> @kandersolar Could you help with the flake8 issue - I can't really tell what it is?

I think the problem is this:
1) To use flake8's `--diff` option, we are piping `git diff` into flake8 via stdin rather than letting flake8 discover and read files itself
2) That means flake8 has to identify .py files by parsing the `git diff` text provided on stdin, which means it wants to decode the stdin text to UTF-8 first
3) The CSV data file has that troublesome trademark character that cannot be decoded as UTF-8.  This character is included in the output of `git diff`
4) Thus flake8 breaks while trying to decode the data passed to its stdin

What to do about it?  Modify the `git diff` to only show output for `.py` files like this, maybe?

```bash
git diff upstream/main -- "*.py" | flake8 --exclude pvlib/version.py --ignore E201,E241,E226,W503,W504 --max-line-length 79 --diff
```

**Comment by AdamRJensen** [comment]
> I think the problem is this:
> 
> 1. To use flake8's `--diff` option, we are piping `git diff` into flake8 via stdin rather than letting flake8 discover and read files itself
> 2. That means flake8 has to identify .py files by parsing the `git diff` text provided on stdin, which means it wants to decode the stdin text to UTF-8 first
> 3. The CSV data file has that troublesome trademark character that cannot be decoded as UTF-8.  This character is included in the output of `git diff`
> 4. Thus flake8 breaks while trying to decode the data passed to its stdin
> 
> What to do about it? Modify the `git diff` to only show output for `.py` files like this, maybe?
> 
> ```shell
> git diff upstream/main -- "*.py" | flake8 --exclude pvlib/version.py --ignore E201,E241,E226,W503,W504 --max-line-length 79 --diff
> ```

This solved the issue. Thanks!

File: .github/workflows/flake8.yml
Comment (kandersolar) [inline .github/workflows/flake8.yml:23]:
```suggestion
        run: git diff upstream/$GITHUB_BASE_REF HEAD -- "*.py" | flake8
```

We should probably keep this part, just in case we ever start sending PRs to branches other than `main`.  Sorry I didn't think to do that in the original suggestion!

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [inline pvlib/iotools/solaranywhere.py:70]:
```suggestion
        Last timestamp of the requested period. If a timezone is not
```

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [inline pvlib/iotools/solaranywhere.py:284]:
```suggestion
    data.index = pd.to_datetime(data['ObservationTime(LST)'],
                                format='%m/%d/%Y %H:%M')
```
Maybe better to use the local standard time column?

File: pvlib/tests/conftest.py
Comment (kandersolar) [inline pvlib/tests/conftest.py:104]: I don't understand why codecov is claiming that line 102 isn't covered.  The CI logs show that the tests are running, so I don't see why this line wouldn't be getting hit.  No action needed; just recording my confusion...

File: pvlib/iotools/solaranywhere.py
Comment (kandersolar) [reply pvlib/iotools/solaranywhere.py:284]: If so, I think the tests will need to be adjusted

**Review by kandersolar — APPROVED** [review]
A few minor comments, but this looks good to me!  Note that I did not actually run any code.

File: pvlib/iotools/solaranywhere.py
Comment (AdamRJensen) [reply pvlib/iotools/solaranywhere.py:284]: I think you are right that LST is preferred. My main point for saying this is that it's much easier for a user to convert to UTC than to convert to LST,

File: pvlib/iotools/solaranywhere.py
Comment (cwhanse) [inline pvlib/iotools/solaranywhere.py:99]:
```suggestion
        where applicable. See :const:`VARIABLE_MAP`.
```

File: pvlib/iotools/solaranywhere.py
Comment (cwhanse) [inline pvlib/iotools/solaranywhere.py:228]:
```suggestion
    additional variables and metadata.
```

Too bad we can't coin this term for substantial metadata, as opposed to thin data.

File: pvlib/iotools/solaranywhere.py
Comment (cwhanse) [inline pvlib/iotools/solaranywhere.py:236]:
```suggestion
        where applicable. See :const:`VARIABLE_MAP`.
```

File: pvlib/tests/iotools/test_solaranywhere.py
Comment (cwhanse) [inline pvlib/tests/iotools/test_solaranywhere.py:18]:
```suggestion
    Users can freely register for an API key."""
```

**Review by cwhanse — APPROVED** [review]
One comment about requirements `start` and `end`, rest are spelling-level nits.

File: docs/sphinx/source/whatsnew/v0.9.2.rst
Comment (AdamRJensen) [reply docs/sphinx/source/whatsnew/v0.9.2.rst:1]: There's no change there so I can't remove the file... I think it's fine leaving it. @kandersolar leave a thumb if you have an opinion

File: pvlib/iotools/solaranywhere.py
Comment (AdamRJensen) [reply pvlib/iotools/solaranywhere.py:161]: I switched to `isinstance`

File: pvlib/iotools/solaranywhere.py
Comment (AdamRJensen) [reply pvlib/iotools/solaranywhere.py:167]: If a user specified either `start` or `end` we would want them to get a useful feedback. Such as `None` cannot be converted to timestamp which is what would happen now. If you use `and` then the error message would first happen much later and be more difficult to interpret.

**Comment by kandersolar** [comment]
Ok, going ahead with the merge.  Thanks @AdamRJensen!

**Comment by KWagnerCPR** [comment]
We greatly appreciate the work on this!

**Pull request merged**
